cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/DefaultSecureStoreService.java

```diff
@@ -0,0 +1,126 @@
+    Principal principal = SecurityRequestContext.toPrincipal();
+    authorizerInstantiator.get().enforce(namespaceId, principal, Action.READ);
+    List<SecureStoreMetadata> metadatas = secureStore.list(namespaceId.getNamespace());
+    List<SecureKeyListEntry> returnList = new ArrayList<>(metadatas.size());
+    for (SecureStoreMetadata metadata : metadatas) {
+      returnList.add(new SecureKeyListEntry(metadata.getName(), metadata.getDescription()));
+    }
+    return returnList;
+  }
+
+  /**
+   * Checks if the user has access to read the secure key and returns the data associated with the key if they do.
+   * @param secureKeyId Id of the key that the user is trying to read.
+   * @return Data associated with the key if the user has read access.
+   * @throws Exception If either the caller does not have READ access to the key or if there was a problem
+   * getting the data from the underlying provider.
+   */
+  @Override
+  public SecureStoreData get(SecureKeyId secureKeyId) throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
+    authorizerInstantiator.get().enforce(secureKeyId, principal, Action.READ);
+    return secureStore.get(secureKeyId.getNamespace(), secureKeyId.getName());
+  }
+
+  /**
+   * Puts the user provided data in the secure store, if the user has write access to the namespace.
+   * @param secureKeyCreateRequest The request containing the data to be stored in the secure store.
+   * @param secureKeyId The Id for the key that needs to be stored.
+   * @throws Exception If either the caller does not have WRITE access to the namespace
+   * or if there was a problem storing the data to the underlying provider
+   * or if the request did not contain the value to be stored.
+   */
+  @Override
+  public synchronized void put(SecureKeyCreateRequest secureKeyCreateRequest, SecureKeyId secureKeyId)
+    throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
+    NamespaceId namespaceId = new NamespaceId(secureKeyId.getNamespace());
+    authorizerInstantiator.get().enforce(namespaceId, principal, Action.WRITE);
+    String description = secureKeyCreateRequest.getDescription();
+    String value = secureKeyCreateRequest.getData();
+    if (Strings.isNullOrEmpty(value)) {
+      throw new BadRequestException("The data field should not be empty. This is the data that will be stored " +
+                                      "securely.");
+    }
+
+    byte[] data = value.getBytes(StandardCharsets.UTF_8);
+    secureStoreManager.put(secureKeyId.getNamespace(), secureKeyId.getName(), data, description,
+                           secureKeyCreateRequest.getProperties());
+  }
+
+  /**
+   * Deletes the key if the user has ADMIN privileges to the key.
+   * @param secureKeyId Id of the key to be deleted.
+   * @throws Exception If either the caller does not have ADMIN access to the key or if there was a problem
```

Comment:
I think its better to add a separate ``@throws UnauthorizedException`` that describes the authorization failure, and also specifies that the failure will happen if both perimeter security and authorization are enabled. Applies to all methods.